File: examples/rust/graph.rs
// Copyright (c) 2026 vectorless developers
// SPDX-License-Identifier: Apache-2.0

//! Document graph example for Vectorless.
//!
//! Demonstrates how to retrieve the cross-document relationship graph
//! after indexing. The graph is automatically rebuilt after each index call,
//! connecting documents that share keywords via Jaccard similarity.
//!
//! # Usage
//!
//! ```bash
//! cargo run --example graph
//! ```

use vectorless::{EngineBuilder, IndexContext};

#[tokio::main]
async fn main() -> vectorless::Result<()> {
    println!("=== Document Graph Example ===\n");

    // 1. Create engine
    let engine = EngineBuilder::new()
        .with_workspace("./workspace_graph_example")
        .build()
        .await
        .map_err(|e: vectorless::BuildError| vectorless::Error::Config(e.to_string()))?;

    // 2. Index documents — graph is rebuilt automatically
    let result = engine
        .index(IndexContext::from_paths(&["./README.md", "./CLAUDE.md"]))
        .await?;

    println!("Indexed {} document(s)", result.items.len());
    for item in &result.items {
        println!("  - {} ({})", item.name, item.doc_id);
    }
    println!();

    // 3. Get the document graph
    match engine.get_graph().await? {
        Some(graph) => {
            println!(
                "Document graph: {} nodes, {} edges",
                graph.node_count(),
                graph.edge_count()
            );

            // Show document nodes
            for doc_id in graph.doc_ids() {
                if let Some(node) = graph.get_node(doc_id) {
                    println!(
                        "  Node: {} — {} keyword(s), top: {:?}",
                        node.title,
                        node.top_keywords.len(),
                        node.top_keywords
                            .iter()
                            .take(3)
                            .map(|kw| &kw.keyword)
                            .collect::<Vec<_>>()
                    );

                    // Show edges (connected documents)
                    let neighbors = graph.get_neighbors(doc_id);
                    if !neighbors.is_empty() {
                        for edge in neighbors {
                            println!(
                                "    → {} (weight={:.2}, jaccard={:.2}, shared={})",
                                edge.target_doc_id,
                                edge.weight,
                                edge.evidence.keyword_jaccard,
                                edge.evidence.shared_keyword_count,
                            );
                        }
                    } else {
                        println!("    (no connections)");
                    }
                }
            }
        }
        None => println!("No graph available (no documents with reasoning index)"),
    }

    // 4. Cleanup
    let docs = engine.list().await?;
    for doc in &docs {
        engine.remove(&doc.id).await?;
    }

    println!("\n=== Done ===");
    Ok(())
}

File: vectorless/types/graph.py
"""Typed Python wrappers for PyO3 graph types."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple


@dataclass(frozen=True)
class WeightedKeyword:
    """A keyword with importance weight."""

    keyword: str
    weight: float

    @classmethod
    def from_rust(cls, item: object) -> WeightedKeyword:
        return cls(keyword=item.keyword, weight=item.weight)


@dataclass(frozen=True)
class EdgeEvidence:
    """Evidence for a graph edge connecting two documents."""

    shared_keyword_count: int = 0
    keyword_jaccard: float = 0.0
    shared_keywords: Tuple[Tuple[str, float], ...] = ()

    @classmethod
    def from_rust(cls, item: object) -> EdgeEvidence:
        keywords = tuple((kw.keyword, kw.weight) for kw in item.shared_keywords)
        return cls(
            shared_keyword_count=item.shared_keyword_count,
            keyword_jaccard=item.keyword_jaccard,
            shared_keywords=keywords,
        )


@dataclass(frozen=True)
class GraphEdge:
    """An edge in the document relationship graph."""

    target_doc_id: str
    weight: float
    evidence: Optional[EdgeEvidence] = None

    @classmethod
    def from_rust(cls, item: object) -> GraphEdge:
        evidence = EdgeEvidence.from_rust(item.evidence) if item.evidence else None
        return cls(
            target_doc_id=item.target_doc_id,
            weight=item.weight,
            evidence=evidence,
        )


@dataclass(frozen=True)
class GraphNode:
    """A document node in the relationship graph."""

    doc_id: str
    title: str
    format: str
    node_count: int
    top_keywords: List[WeightedKeyword] = field(default_factory=list)

    @classmethod
    def from_rust(cls, item: object) -> GraphNode:
        keywords = [WeightedKeyword.from_rust(kw) for kw in item.top_keywords]
        return cls(
            doc_id=item.doc_id,
            title=item.title,
            format=item.format,
            node_count=item.node_count,
            top_keywords=keywords,
        )


@dataclass
class DocumentGraphWrapper:
    """Typed wrapper around the cross-document relationship graph."""

    _inner: Any

    @classmethod
    def from_rust(cls, graph: object) -> DocumentGraphWrapper:
        return cls(_inner=graph)

    def node_count(self) -> int:
        return self._inner.node_count()

    def edge_count(self) -> int:
        return self._inner.edge_count()

    def get_node(self, doc_id: str) -> Optional[GraphNode]:
        node = self._inner.get_node(doc_id)
        return GraphNode.from_rust(node) if node is not None else None

    def get_neighbors(self, doc_id: str) -> List[GraphEdge]:
        neighbors = self._inner.get_neighbors(doc_id)
        return [GraphEdge.from_rust(e) for e in neighbors]

    def doc_ids(self) -> List[str]:
        return list(self._inner.doc_ids())

    def is_empty(self) -> bool:
        return self._inner.is_empty()
